### PR TITLE
Switch to graphql-config

### PR DESCRIPTION
**Pull request by RomanGotsiy**
hey @asiandrummer ,
here is the PR to switch from `graphql-language-services-config` to `graphql-config`.
I split it into 2 commits. First one removes `graphql-language-services-config` folder and the second one adds support for graphqlconfig

All the tests are passing and flow doesn't show any errors. Although it definitely requires looking from someone who knows the codebase

### CODE REVIEW

**Comment by asiandrummer** [comment]
Checking this out. Thanks!

**Comment by asiandrummer** [comment]
@RomanGotsiy This looks amazing! I think I'll be able to merge this, but just in case am doing a super granular read for safety. One high level feedback - would it be possible to PR out the removal of `graphql-language-service-config` separately?

**Comment by RomanGotsiy** [comment]
@asiandrummer 
> I think I'll be able to merge this, but just in case am doing a super granular read for safety
Definitely! Would be great if you also can test it with Nuclide if possible as there may be some issues in `graphql-config`.

> would it be possible to PR out the removal of graphql-language-service-config separately?
I think that no. Other packages depend on `graphql-language-service-config` so tests won't pass and flow will complain.

there is probably a need to add some function for detecting and converting old config format to new just to support old configs. Do you think this is required? Do you want me to add this?

**Comment by asiandrummer** [comment]
> One more thing: there is probably a need to add some function for detecting and converting old config format to new just to support old configs. Do you think this is required? Do you want me to add this?

I just realized that you've changed `includeDirs` and `excludeDirs` to `include` and `exclude`. I probably missed this while checking out the PR from `graphql-config` repo - is it too late to revert back to `includeDirs` and `excludeDirs`? I think `include` and `exclude` might be too generic to specify what it wants to accomplish.

Since I'm late to suggest this change, if you've already made a lot of PRs to apply this to other repos, I'm fine with writing a function to enable backward-compatibility.

File: packages/graphql-language-service-interface/src/GraphQLLanguageService.js [hunk 1]
```diff
@@ -54,18 +55,14 @@ export class GraphQLLanguageService {
...
+    const projectConfig = this._graphQLConfig.getConfigForFile(uri);
     // If there's a matching config, proceed to prepare to run validation
     let schema;
     let customRules;
-    if (this._graphQLConfig.getSchemaPath(appName)) {
-      schema = await this._graphQLCache.getSchema(
-        this._graphQLConfig.getSchemaPath(appName),
-        appName,
-      );
+    if (projectConfig.schemaPath) {
```
Comment (asiandrummer) [inline packages/graphql-language-service-interface/src/GraphQLLanguageService.js]: I like this getter stuff ;)

File: packages/graphql-language-service-interface/src/GraphQLLanguageService.js [hunk 2]
```diff
@@ -103,10 +98,7 @@ export class GraphQLLanguageService {
     const appName = this._graphQLConfig.getAppConfigNameByFilePath(filePath);
```
Comment (asiandrummer) [inline packages/graphql-language-service-interface/src/GraphQLLanguageService.js]: Missed here I think

Comment (asiandrummer) [inline packages/graphql-language-service-interface/src/GraphQLLanguageService.js]: Needs to be:
```
const projectConfig = this._graphQLConfig.getConfigForFile(filePath);
```
and so on, similar to `getDiagnostics`

File: packages/graphql-language-service-interface/src/__tests__/GraphQLLanguageService-test.js [hunk 3]
```diff
@@ -10,18 +10,19 @@
...
 import {beforeEach, describe, it} from 'mocha';
+import {join} from 'path';
 
-import {GraphQLConfig} from 'graphql-language-service-config';
+import {GraphQLConfig} from 'graphql-config';
 import {GraphQLLanguageService} from '../GraphQLLanguageService';
 
 const MOCK_CONFIG = {
-  inputDirs: ['./queries'],
+  include: ['./queries/**'],
```
Comment (asiandrummer) [inline packages/graphql-language-service-interface/src/__tests__/GraphQLLanguageService-test.js]: I love it

File: packages/graphql-language-service-server/src/GraphQLCache.js [hunk 4]
```diff
@@ -155,20 +149,18 @@ export class GraphQLCache {
...
 
   getFragmentDefinitions = async (
-    graphQLConfig: GraphQLConfigInterface,
-    appName: ?string,
+    projectConfig: GraphQLProjectConfig,
   ): Promise<Map<string, FragmentInfo>> => {
     // This function may be called from other classes.
     // If then, check the cache first.
-    const rootDir = graphQLConfig.getRootDir();
+    const rootDir = projectConfig.configDir;
```
Comment (asiandrummer) [inline packages/graphql-language-service-server/src/GraphQLCache.js]: I think you forgot to do a getter for `configDir` in https://github.com/graphcool/graphql-config/blob/master/src/GraphQLProjectConfig.ts, which makes me feel a little uncomfortable. If possible could we be consistent and use the proper getter? Not really a blocker though

File: packages/graphql-language-service-server/src/GraphQLCache.js [hunk 5]
```diff
@@ -178,10 +170,7 @@ export class GraphQLCache {
         mtime: fileInfo.mtime,
         // Filter any files with path starting with ExcludeDirs
       }))
-      .filter(fileInfo =>
-        excludeDirs.every(
-          exclude => !fileInfo.filePath.startsWith(path.join(rootDir, exclude)),
-        ));
+      .filter(fileInfo => projectConfig.includesFile(fileInfo.filePath));
```
Comment (asiandrummer) [inline packages/graphql-language-service-server/src/GraphQLCache.js]: This is also a bug fix I think - thanks for that

File: packages/graphql-language-service-server/src/GraphQLCache.js [hunk 6]
```diff
@@ -387,53 +372,40 @@ export class GraphQLCache {
...
-        resolve(
-          content +
-            '\n' +
-            this._graphQLConfig.getCustomDirectives(appName).join('\n'),
-        );
-      }));
-
-    const schemaFileExt = path.extname(schemaPath);
-    let schema;
+    let schemaSDL;
```
Comment (asiandrummer) [inline packages/graphql-language-service-server/src/GraphQLCache.js]: I'm worried of the performance of this approach to generate GraphQL schema. From what I understand, this approach will read the file, build `GraphQLSchema` object, print the schema in SDL format, and build it again. Last time I checked, `buildSchema()` takes around several seconds with 4~5k types available, and with `extendSchema()` being slow as well, this is definitely a hit with the performance.

I guess this is one of the ways to supprot `.json` format, which previously had bugs with my implementation to append the custom directives in SDL format to content string. One easy performance optimization we can make is to optionally perform custom directive extension when `customDirective` extension entry is available, with `extendSchema`, not `buildSchema`.

**Comment by asiandrummer** [comment]
Really amazing work @RomanGotsiy!

File: packages/graphql-language-service-interface/src/GraphQLLanguageService.js (hunk 2 above)
Comment (RomanGotsiy) [inline packages/graphql-language-service-interface/src/GraphQLLanguageService.js]: Good catch! Fixed.
Probably there is need to add some test-case that covers this piece of code

File: packages/graphql-language-service-server/src/GraphQLCache.js (hunk 4 above)
Comment (RomanGotsiy) [inline packages/graphql-language-service-server/src/GraphQLCache.js]: Agree, added getter for configDir.

File: packages/graphql-language-service-server/src/GraphQLCache.js (hunk 6 above)
Comment (RomanGotsiy) [inline packages/graphql-language-service-server/src/GraphQLCache.js]: Great idea! I refactored code to use `extendSchema` when `customDirective` extension entry is available

Just one thing. It appeared that `extendSchema` loses directive description. I've already opened a PR ([#961](https://github.com/graphql/graphql-js/pull/961)) on `graphql-js` to fix that but had to change `""` to `undefined` in [one test case](https://github.com/graphql/graphql-language-service/pull/130/commits/64eca9c6e621b843aec61cde68a5cdce64d64a20#diff-a1bf6c6aa3f36f3fb1872c4fd4628d61R54) for the time being (until my PR is merged)

**Comment by RomanGotsiy** [comment]
Thanks, @asiandrummer!
So far I have fixed all the issues you've spotted!

Let me know if there is anything else

**Comment by asiandrummer** [comment]
lgtm, cc @wincent for visibility
@RomanGotsiy what are you thoughts on the comment I made above about `include` and `exclude`?

>I just realized that you've changed includeDirs and excludeDirs to include and exclude. I probably missed this while checking out the PR from graphql-config repo - is it too late to revert back to includeDirs and excludeDirs? I think include and exclude might be too generic to specify what it wants to accomplish.

> Since I'm late to suggest this change, if you've already made a lot of PRs to apply this to other repos, I'm fine with writing a function to enable backward-compatibility.

**Comment by RomanGotsiy** [comment]
>  what are you thoughts on the comment I made above about include and exclude?

oops, missed that.

> I think include and exclude might be too generic to specify what it wants to accomplish.

I partly agree. But at the same time globs are much more powerful. Let's say I want to exclude all the `__test__` directories down to the directories tree. Using glob I can set:

```
"exclude": "**/__test__/**"
```

while using `excludeDirs` I will have to specify each folder + each time I create a new one I have to add it to the config.

On the other hand globs are harder to work with. + there is no simple and reliable way to convert glob to list of dirs. Does fb-watchman support globs?

**Comment by asiandrummer** [comment]
@RomanGotsiy sorry - I meant only the name, not the functionality :D I love the glob support!
`includeDirs` -> `include` and `excludeDirs` -> `exclude` might have been too aggressive of a change, at least for `graphql-language-service`. Again I do love the glob support, but could we just change the names back to `includeDirs` and `excludeDirs`?

**Comment by RomanGotsiy** [comment]
But if a user specifies glob `"test/**/*.graphql"` which is basically a file glob. Then `includeDirs` name doesn't work.
But I see now why `include` may be a too generic name. How about `includePattern/excludePattern` or `includeGlob/excludeGlob` ?

**Comment by asiandrummer** [comment]
Or maybe we could name them `includes/exclude` - sounds more functional.

**Comment by RomanGotsiy** [comment]
Yes! I love this idea 👍  I will update specification/implementation and this PR to use `includes` and `excludes`.

btw, You probably missed one `s` (`includes/exclude` <- here) in the previous comment :)

**Comment by RomanGotsiy** [comment]
Hey @asiandrummer,

I've updated graphql-config and this PR to use `includes`/`excludes`

**Comment by schickling** [comment]
Any update on this @asiandrummer? :)

**Comment by asiandrummer** [comment]
Sorry I've been away - thanks for the contribution!

**Pull request merged**